astropy_helpers/commands/build_ext.py

```diff
@@ -1,329 +1,180 @@
-                with open(cython_py, 'w') as f:
-                    f.write('# Generated file; do not modify\n')
-                    f.write('cython_version = {0!r}\n'.format(self.uses_cython))
-
-                if os.path.isdir(self.build_lib):
-                    # The build/lib directory may not exist if the build_py
-                    # command was not previously run, which may sometimes be
-                    # the case
-                    self.copy_file(cython_py,
-                                   os.path.join(self.build_lib, cython_py),
-                                   preserve_mode=False)
-
-                invalidate_caches()
-
-        def _check_cython_sources(self, extension):
-            """
-            Where relevant, make sure that the .c files associated with .pyx
-            modules are present (if building without Cython installed).
-            """
-
-            # Determine the compiler we'll be using
-            if self.compiler is None:
-                compiler = get_default_compiler()
-            else:
-                compiler = self.compiler
+    uses_cython = False
+    force_rebuild = False
+
+    def finalize_options(self):
 
-            # Replace .pyx with C-equivalents, unless c files are missing
-            for jdx, src in enumerate(extension.sources):
-                base, ext = os.path.splitext(src)
-                pyxfn = base + '.pyx'
-                cfn = base + '.c'
-                cppfn = base + '.cpp'
+        # First let's find the package folder, then we can check if the
+        # version and cython_version are accessible
+        self.package_dir = get_main_package_directory(self.distribution)
 
-                if not os.path.isfile(pyxfn):
-                    continue
+        version = import_file(os.path.join(self.package_dir, 'version.py'),
+                              name='version').version
+        self.is_release = 'dev' not in version
 
-                if self.uses_cython:
-                    extension.sources[jdx] = pyxfn
+        try:
+            self.previous_cython_version = import_file(os.path.join(self.package_dir,
+                                                                    'cython_version.py'),
+                                                       name='cython_version').cython_version
+        except (FileNotFoundError, ImportError):
+            self.previous_cython_version = 'unknown'
+
+        self.uses_cython = should_build_with_cython(self.previous_cython_version, self.is_release)
+
+        # Add a copy of the _compiler.so module as well, but only if there
+        # are in fact C modules to compile (otherwise there's no reason to
+        # include a record of the compiler used). Note that self.extensions
+        # may not be set yet, but self.distribution.ext_modules is where any
+        # extension modules passed to setup() can be found
+        extensions = self.distribution.ext_modules
```

Comment:
Maybe this should go *above* the comment block, since the comment actually applies for the below part more directly?

Reply:
The ``extensions = ...`` is actually part of the block below, but I've removed the empty line below because I agree it looked confusing before. Now it's all one block of code it should be clearer